service/telemetry/internal/migration/v0.3.0.go

```diff
@@ -0,0 +1,174 @@
+
+	// Sampling sets a sampling policy.
+	// Default:
+	// 		sampling:
+	//	   		enabled: true
+	//	   		tick: 10s
+	//	   		initial: 10
+	//	   		thereafter: 100
+	// Sampling can be disabled by setting 'enabled' to false
+	Sampling *LogsSamplingConfig `mapstructure:"sampling"`
+
+	// OutputPaths is a list of URLs or file paths to write logging output to.
+	// The URLs could only be with "file" schema or without schema.
+	// The URLs with "file" schema must be an absolute path.
+	// The URLs without schema are treated as local file paths.
+	// "stdout" and "stderr" are interpreted as os.Stdout and os.Stderr.
+	// see details at Open in zap/writer.go.
+	// (default = ["stderr"])
+	OutputPaths []string `mapstructure:"output_paths"`
+
+	// ErrorOutputPaths is a list of URLs or file paths to write zap internal logger errors to.
+	// The URLs could only be with "file" schema or without schema.
+	// The URLs with "file" schema must use absolute paths.
+	// The URLs without schema are treated as local file paths.
+	// "stdout" and "stderr" are interpreted as os.Stdout and os.Stderr.
+	// see details at Open in zap/writer.go.
+	//
+	// Note that this setting only affects the zap internal logger errors.
+	// (default = ["stderr"])
+	ErrorOutputPaths []string `mapstructure:"error_output_paths"`
+
+	// InitialFields is a collection of fields to add to the root logger.
+	// Example:
+	//
+	// 		initial_fields:
+	//	   		foo: "bar"
+	//
+	// By default, there is no initial field.
+	InitialFields map[string]any `mapstructure:"initial_fields"`
+
+	// Processors allow configuration of log record processors to emit logs to
+	// any number of suported backends.
+	Processors []config.LogRecordProcessor `mapstructure:"processors"`
+}
+
+// LogsSamplingConfig sets a sampling strategy for the logger. Sampling caps the
+// global CPU and I/O load that logging puts on your process while attempting
+// to preserve a representative subset of your logs.
+type LogsSamplingConfig struct {
+	// Enabled enable sampling logging
+	Enabled bool `mapstructure:"enabled"`
+	// Tick represents the interval in seconds that the logger apply each sampling.
+	Tick time.Duration `mapstructure:"tick"`
+	// Initial represents the first M messages logged each Tick.
+	Initial int `mapstructure:"initial"`
+	// Thereafter represents the sampling rate, every Nth message will be sampled after Initial messages are logged during each Tick.
+	// If Thereafter is zero, the logger will drop all the messages after the Initial each Tick.
+	Thereafter int `mapstructure:"thereafter"`
+}
+
+func (c *LogsConfigV030) Unmarshal(conf *confmap.Conf) error {
+	if err := conf.Unmarshal(c); err != nil {
+		var v02 logsConfigV020
+		// try unmarshaling using v0.2.0 struct
+		if e := conf.Unmarshal(&v02); e != nil {
+			// error could not be resolved through backwards
+			// compatibility attempts
+			return err
+		}
+		// TODO: add a warning log to tell users to migrate their config
```

Comment:
I guess it might be complicated to pass a logger here, but it's pretty important. I think we should come up with a solution by the time we release this change

Reply:
The way we do this on the Datadog exporter is to have [a private field](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/adef54ee022ddd503bf78afdfb2a01b599853579/pkg/datadog/config/config.go#L104) that is [written to during unmarshaling](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/adef54ee022ddd503bf78afdfb2a01b599853579/pkg/datadog/config/config.go#L267) and then [a method to log warnings](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/adef54ee022ddd503bf78afdfb2a01b599853579/pkg/datadog/config/config.go#L107-L112).

I don't love it, but I haven't been able to come up with anything better